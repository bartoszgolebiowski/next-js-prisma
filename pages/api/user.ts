import { PrismaClient } from "@prisma/client";
import { NextApiRequest, NextApiResponse } from "next";

import {
  getAllUsers,
  createUser,
  createUserValidator,
} from "../../src/db/user";

const prisma = new PrismaClient();

export default async (req: NextApiRequest, res: NextApiResponse) => {
  if (req.method === "GET") {
    const users = await getAllUsers(prisma);
    res.statusCode = 200;
    res.json({ users });
  }
  if (req.method === "POST") {
    try {
      const user = req.body;
      createUserValidator(user);
      const persistedUser = await createUser(prisma, user);
      res.statusCode = 201;
      res.json({ user: persistedUser });
    } catch ({ errors, name, ...rest }) {
      res.statusCode = 400;
      res.json({ errors, name });
    }
  }
};
