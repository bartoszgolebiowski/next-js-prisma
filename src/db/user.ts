import { Prisma, PrismaClient, User } from "@prisma/client";

type UserPersonalData = Prisma.UserGetPayload<{
  select: { email: true; name: true };
}>;

export const getAllUsers = async (client: PrismaClient): Promise<User[]> => {
  return client.user.findMany();
};

export const createUser = async (
  client: PrismaClient,
  user: UserPersonalData
): Promise<User> => {
  const userPersisted = await client.user.create({
    data: { ...user },
  });
  return userPersisted;
};
