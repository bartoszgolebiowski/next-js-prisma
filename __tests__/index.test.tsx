import { User } from "@prisma/client";
import { render, RenderResult, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import axios from "axios";

import Home, { getStaticProps } from "../pages";
import * as userService from "../src/db/user";

import { createuserBody2, getAllUsersSample } from "../test/samples/db";

jest.mock("@prisma/client", () => ({
  PrismaClient: jest.fn(),
}));

describe("/", () => {
  let getAllUsers: jest.SpyInstance<Promise<User[]>>;
  let axiosSpy: any;
  let baseElement: Element;

  beforeEach(async () => {
    getAllUsers = jest.spyOn(userService, "getAllUsers");
    axiosSpy = jest.spyOn(axios, "post");

    getAllUsers.mockResolvedValue(getAllUsersSample);
    const { props } = await getStaticProps();

    const container = render(<Home {...props} />);
    baseElement = container.baseElement;
  });

  it("initla render", async () => {
    expect(baseElement).toMatchSnapshot();
  });

  it("create user", async () => {
    userEvent.type(screen.getAllByLabelText(/name:/i)[0], createuserBody2.name);
    userEvent.type(screen.getByLabelText(/email/i), createuserBody2.email);
    userEvent.type(screen.getByLabelText(/surname:/i), createuserBody2.surname);
    userEvent.click(screen.getByText(/submit/i));
    expect(axiosSpy).toBeCalledWith("/api/user", createuserBody2);
  });
});
