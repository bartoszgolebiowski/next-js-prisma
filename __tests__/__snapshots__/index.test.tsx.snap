// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/ initla render 1`] = `
<body>
  <div>
    <form>
      <label>
        Name:
        <input
          name="name"
          required=""
        />
      </label>
      <label>
        Surname:
        <input
          name="surname"
        />
      </label>
      <label>
        Email:
        <input
          name="email"
          required=""
          type="email"
        />
      </label>
      <button
        type="submit"
      >
        Submit
      </button>
    </form>
    <div>
      [{"id":1,"email":"test1@gmail.com","name":"name1","surname":"surname1"},{"id":2,"email":"test2@gmail.com","name":"name2","surname":"surname2"}]
    </div>
  </div>
</body>
`;
